fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/simplification/Simplification.java

```diff
@@ -96,39 +105,64 @@ public static Value simplify(@Nonnull final Value root,
             final var executionResult = executeRuleSetIteratively(isRoot ? current : root,
                     current,
                     ruleSet,
-                    (rule, r, c, plannerBindings) -> new ValueSimplificationRuleCall(rule, r, c, plannerBindings,
-                            aliasMap, constantAliases),
-                    Iterables::getOnlyElement);
+                    (rule, r, c, plannerBindings) ->
+                            new ValueSimplificationRuleCall(rule, r, c, evaluationContext, plannerBindings, aliasMap,
+                                    constantAliases, constraintsMap::get),
+                    (results, queryPlanConstraint) ->
+                            onResultsFunctionForSimplification(constraintsMap, results, queryPlanConstraint));
             Verify.verify(!executionResult.shouldReExplore());
             return executionResult.getBase();
         }).orElseThrow(() -> new RecordCoreException("expected a mapped tree"));
+        return Constrained.ofConstrainedObject(simplifiedValue, collectAndComposeConstraints(simplifiedValue, constraintsMap));
+    }
+
+    @Nonnull
+    private static <BASE extends TreeLike<BASE>> QueryPlanConstraint collectAndComposeConstraints(@Nonnull final BASE root,
+                                                                                                  @Nonnull final Map<BASE, QueryPlanConstraint> constraintMap) {
+        return root.postOrderStream()
+                .flatMap(current -> {
+                    final var constraint = constraintMap.get(current);
+                    return Stream.ofNullable(constraint);
+                })
+                .filter(QueryPlanConstraint::isConstrained)
+                .reduce(QueryPlanConstraint::compose)
+                .orElse(QueryPlanConstraint.noConstraint());
     }
 
     /**
      * Main function that simplifies the given value using the {@link AbstractValueRuleSet} passed in.
      * @param current the {@link Value} to be simplified
+     * @param evaluationContext the evaluation context
      * @param aliasMap an alias map of equalities
      * @param constantAliases a set of aliases that are considered to be constant
      * @param ruleSet the rule set used to simplify the {@link Value} that is passed in
      * @return a new simplified {@link Value} of {@code root}
```

Comment:
While you're at it, can you please fix the documentation of the returned `Value`(s)?

Reply:
done.